engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+			return fmt.Errorf("serverrescuekeys differs")
+		}
+	}
+	return nil
+}
+
+// cmpServerSpecs compares the server specifications between the virtual
+// HetznerVMRes resource and its corresponding Hetzner instance. Returns true if
+// ServerType, Datacenter and Image match. Returns an error is the server is
+// unavailable (`absent`).
+func (obj HetznerVMRes) cmpServerSpecs() (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("cmpServerSpecs()")
+	}
+	if obj.server == nil {
+		return false, fmt.Errorf("server is unavailable")
+	}
+	if obj.ServerType != obj.server.ServerType.Name {
+		return false, nil
+	}
+	if obj.Datacenter != obj.server.Datacenter.Name {
+		return false, nil
+	}
+	if obj.Image != obj.server.Image.Name {
+		return false, nil
+	}
+	return true, nil
+}
+
+// createServer checks if the servername is not already in use, builds the
+// serverconfig in hcloud-go format from resource params, requests a server
+// creation with that configuration, and waits until the creation is confirmed.
+// Errors occur when the server exists already, the client fails, or the wait
+// step exits early (context cancellation, client failure or timeout).
+// TODO: add createServer permission field
+func (obj *HetznerVMRes) createServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("createServer()")
+	}
+	if obj.server != nil {
+		return fmt.Errorf("server already exists")
+	}
+	if err := obj.getServerConfig(ctx); err != nil {
+		return errwrap.Wrapf(err, "getServerConfig failed")
+	}
+	if _, _, err := obj.client.Server.Create(ctx, obj.serverconfig); err != nil {
+		return errwrap.Wrapf(err, "client.server.create failed")
+	}
+	if err := obj.waitFor(ctx, serverExists); err != nil {
+		return errwrap.Wrapf(err, "waitFor(serverExists) exited early")
+	}
+	return nil
+}
+
+// deleteServer checks if the server is available from the client, requests a
+// server deletion from the API, waits for confirmation, then returns. An error
+// is thrown when the server is already unavailable or something fails. Context
+// cancellation allows a clean exit when needed.
+// NOTE: this is a destructive operation. Check permissions before applying!
+// TODO: move the AllowDeletion policy inside this function for safety?
+func (obj *HetznerVMRes) deleteServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("deleteServer()")
+	}
+	if obj.server == nil {
+		return fmt.Errorf("server is already unavailable")
+	}
+	if _, err := obj.client.Server.Delete(ctx, obj.server); err != nil {
+		return errwrap.Wrapf(err, "client.server.delete failed")
+	}
+	if err := obj.waitFor(ctx, serverAbsent); err != nil {
+		return errwrap.Wrapf(err, "waitFor(serverAbsent) exited early")
+	}
+	return nil
+}
+
+// rebuildServer deletes the current server instance and creates a new one, in
+// accordance with the provided resource specifications.
+// NOTE: this is a destructive operation, don't call without AllowDeletion.
```

Comment:
Then why aren't we checking AllowDeletion in _here_ ??

Reply:
Agreed, will move those protections into these functions directly.